sunpy/coordinates/utils.py

```diff
@@ -265,3 +265,74 @@ def coordinates(self, points=None):
+    top_right : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The top_right coordinate of the rectangle. Can only be omitted if
+        ``bottom_left`` has shape ``(2,)``.
+        If in a different frame than ``bottom_left`` and all required metadata for frame conversion is present,
+        ``top_right`` will be transformed to ``bottom_left`` frame.
+        Otherwise a ``ValueError`` will be incurred.
+    width : `~astropy.units.Quantity`
+        The width of the rectangle.
+        Must be omitted if ``bottom_left`` has shape ``(2,)`` or ``top_right`` is provided.
+    height : `~astropy.units.Quantity`
+        The height of the rectangle.
+        Must be omitted if ``bottom_left`` has shape ``(2,)`` or ``top_right`` is provided.
+
+    Returns
+    -------
+    `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The bottom left coordinate of the rectangular region of interest in spherical world coordinates.
+    `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`
+        The top right coordinate of the rectangular region of interest in spherical world coordinates.
+    """
+    if not (hasattr(bottom_left, 'transform_to') and
+            hasattr(bottom_left, 'shape') and
+            hasattr(bottom_left, 'spherical')):
+        raise TypeError("Invalid input, bottom_left must be of type SkyCoord or BaseCoordinateFrame")
+
+    if (top_right is not None and (hasattr(top_right, 'transform_to') and
+                                   hasattr(top_right, 'shape') and
+                                   hasattr(top_right, 'spherical'))):
+        raise TypeError("Invalid input, top_right if passed, must be of type SkyCoord or BaseCoordinateFrame")
+
+    if bottom_left.shape == (2,) and any((x is not None for x in (width, height, top_right))):
+        raise ValueError("Invalid input, if bottom_left.shape == (2,) "
+                         "other parameters should not be passed.")
+
+    if all(x is not None for x in (width, height, top_right):
+        raise ValueError("Invalid input, width, height and top_right "
+                         "parameters should not be passed simultaneously.")
+
+    if top_right is None and bottom_left.shape != (2,) and not all((width, height)):
+        raise ValueError("Invalid input, either bottom_left and top_right "
+                         "or bottom_left and height and width should be provided.")
+
+    if bottom_left.shape == (2,):
+        top_right = bottom_left[1]
+        bottom_left = bottom_left[0]
+
+    elif top_right is not None:
+        top_right = top_right.transform_to(bottom_left)
+
+    else:
+        if isinstance(bottom_left, BaseCoordinateFrame):
+            base_frame = type(bottom_left)
+            top_right = base_frame(bottom_left.spherical.lon + width,
+                                   bottom_left.spherical.lat + height)
```

Comment:
This needs to be a `bottom_left.realize_frame(...)` call because otherwise you're losing all frame attributes.

Reply:
Does it matter here that the `bottom_left` coord could be a SkyCoord with some stored attrrs and we end up creating a `top_right` which is (currently a frame) but even if we cast to `SkyCoord` it wouldn't have the same history as the `bottom_left`.

Reply:
What was previously there was fine structurally: two code paths depending on whether `bottom_left` was a frame or a SkyCoord.  The frame path needed to call `realize_frame()`, while the SkyCoord path could just use the SkyCoord constructor to indirectly call `realize_frame()` for you.

If the preference is to have a largely single path, you should use the SkyCoord constructor to create a `top_right` as a SkyCoord (which would look the same regardless of the type of `bottom_left`), and then "downgrade" it to a frame if `bottom_left` was actually just a frame.  So, no explicit call to `realize_frame()`.

Reply:
I agree that a comment about the type change would be useful. I am :+1: on the implementation as-is now.